sklearn/metrics/cluster/_unsupervised.py

```diff
@@ -131,15 +132,24 @@ def _silhouette_reduce(D_chunk, start, labels, label_freqs):
     label_freqs : array-like
         Distribution of cluster labels in ``labels``.
     """
+    n_chunk_samples = D_chunk.shape[0]
     # accumulate distances from each sample to each cluster
-    clust_dists = np.zeros((len(D_chunk), len(label_freqs)),
+    clust_dists = np.zeros((n_chunk_samples, len(label_freqs)),
                            dtype=D_chunk.dtype)
-    for i in range(len(D_chunk)):
-        clust_dists[i] += np.bincount(labels, weights=D_chunk[i],
+    for i in range(n_chunk_samples):
+        if issparse(D_chunk):
+            current_chunk = D_chunk.getrow(i)
+            sample_weights = current_chunk.data
+            sample_labels = np.take(labels, current_chunk.indices)
```

Comment:
This looks better.

I think we would need a test to make sure that `sample_weights` work as expected. A simple test would be to run `silhouette_samples` on a sparse dataset and compare it to the result with the same dataset but densified.

Reply:
@thomasjpfan I have added a test for this